README
Qore ssh2 module

INTRODUCTION
------------
This module provides access to ssh2 sessions and the sftp protocol via libssh2
(http://www.libssh2.org) in the Qore programming language.

This version requires qore 0.8.5+ as it requires qpp to build (qpp is the Qore
Pre-Processor)

LICENSE
-------
This module is released under a choice of two licenses: LGPL 2.1 and MIT, see
COPYING.LGPL and COPYING.MIT for more information

BINARIES
--------
For Linux RPMs for many distributions, see the opensuse build service:
    http://download.opensuse.org/repositories/devel:/languages:/misc/

For OS/X, use macports: http://www.macports.org
    sudo port selfupdate
    sudo port -vu install qore-ssh2-module


BUILDING
--------
Requires qore 0.8.6+ and libssh2 1.1+ to build and run

to configure the build, execute
        configure --disable-debug

if the qore library cannot be found, then you can use:
        configure --disable-debug --with-qore=<dir>

if libssh2 cannot be found, then you can use:
        configure --disable-debug --with-libssh2-dir=<dir>

The qore binary also needs to be in the path so configure can determine the
module directory

Then execute

make && make install

(or 'make && sudo make install' as needed)

please direct any questions to:
david@qore.org

or you can post in the qore forums at
        http://qore.org/forums


ChangeLog
2013-08-16 08:36  david_nichols

	* RELEASE-NOTES, configure.ac, docs/mainpage.doxygen.tmpl,
	  qore-ssh2-module.spec: updated version to 0.9.9

2013-08-15 18:41  david_nichols

	* RELEASE-NOTES, docs/mainpage.doxygen.tmpl, src/SFTPClient.cpp,
	  src/SSH2Client.cpp, src/SSH2Client.h: - when errors occur in the
	  SFTPClient class, the connection is marked as closed if the
	  connection has been terminated
	  - made cwd persistent across connections to better deal with
	  implicit disconnections and reconnections

2013-08-15 16:15  david_nichols

	* RELEASE-NOTES, docs/mainpage.doxygen.tmpl, src/QC_SFTPClient.qpp,
	  src/QC_SSH2Base.qpp, src/SFTPClient.cpp, src/SFTPClient.h: -
	  implemented implicit connections for the SFTPClient class
	  - implemented the SFTPClient::listFull() method
	  - implemented the SSH2Base::connected() method

2013-08-05 10:27  david_nichols

	* test/ssh2-test.q: cosmetic fix to usage text

2013-08-05 10:24  david_nichols

	* RELEASE-NOTES, configure.ac, docs/mainpage.doxygen.tmpl,
	  qore-ssh2-module.spec, src/SSH2Client.cpp, src/SSH2Client.h:
	  updated module version to 0.9.8.1 and made additional fixes to
	  non-blocking I/O in ssh2 session initialization

2013-08-03 17:14  david_nichols

	* src/ssh2-module.cpp: compile fixes for older versions of libssh2

2013-08-03 16:06  david_nichols

	* ChangeLog: updated ChangeLog for release

2013-07-04 11:53  subik

	* docs/mainpage.doxygen.tmpl, src/QC_SFTPClient.qpp,
	  src/SFTPClient.cpp, src/SFTPClient.h, test/ssh2-test.q:
	  sftpclient: getTextFile optional encoding

2013-06-03 17:40  david_nichols

	* src/ssh2-module.cpp: fixed for compiling with older libssh2
	  versions

2013-06-03 17:29  david_nichols

	* src/SFTPClient.cpp: minor error msg fixes

2013-06-03 17:09  david_nichols

	* ChangeLog, RELEASE-NOTES, configure.ac,
	  docs/mainpage.doxygen.tmpl, src/QC_SFTPClient.qpp,
	  src/QC_SSH2Base.qpp, src/QC_SSH2Client.qpp, src/SFTPClient.cpp,
	  src/SFTPClient.h, src/SSH2Channel.cpp, src/SSH2Client.cpp,
	  src/SSH2Client.h, src/ssh2-module.cpp, src/ssh2-module.h,
	  test/ssh2-test.q: implemented non-blocking I/O for all sftpclient
	  methods and disconnections, even favoring memory leaks over
	  shutdown/disconnection calls that do not return in the timeout
	  period

2013-05-21 13:11  david_nichols

	* RELEASE-NOTES, docs/mainpage.doxygen.tmpl, src/QC_SSH2Base.qpp:
	  added a default timeout to the connect method

2013-03-12 13:06  david_nichols

	* RELEASE-NOTES, docs/mainpage.doxygen.tmpl, src/SFTPClient.cpp,
	  src/SFTPClient.h: updated to send the same mode as retrieved from
	  the stat() call

2013-03-12 12:34  david_nichols

	* RELEASE-NOTES, configure.ac, docs/mainpage.doxygen.tmpl,
	  qore-ssh2-module.spec, src/SFTPClient.cpp, src/SFTPClient.h:
	  updated to v0.9.8; send mode 0644 when opening files read-only in
	  case broken servers will erroneously set the remote file's mode
	  to 0 and then the file will not be readable; updated docs

2013-02-27 10:10  david_nichols

	* README-BINARIES: updated binary location (and changelog for
	  pgsql)

2012-11-21 08:17  david_nichols

	* configure.ac: use the latest module API for the ssh2 module

2012-10-28 19:04  david_nichols

	* ChangeLog, README, RELEASE-NOTES: updated for 0.9.7 release

2012-10-22 07:05  david_nichols

	* src/SSH2Channel.h: minor correction in a header (for a
	  declaration that's not used anyway)

2012-09-28 16:19  david_nichols

	* configure.ac: applied reini urban's multi-arch patches for debian
	  to all modules

2012-09-20 13:48  david_nichols

	* src/QC_SSH2Base.qpp, src/SSH2Client.cpp, src/SSH2Client.h: * use
	  std::string for connection parameters instead of dynamically
	  managed char*s
	  * throw an exception if trying to use key files where one or both
	  of the key files do not exist or are not readable

2012-09-05 09:24  david_nichols

	* docs/doxyfile.tmpl: disable the collaboration graph because it's
	  redundant

2012-08-19 06:58  david_nichols

	* src/QC_SFTPClient.qpp: fixed a compile warning

2012-08-10 13:56  david_nichols

	* test/ssh2-test.q: allow setting the private key for key-based
	  authentication

2012-07-17 11:51  david_nichols

	* src/QC_SFTPClient.qpp, src/QC_SSH2Client.qpp: fixed class
	  inheritance in dox

2012-06-25 08:51  david_nichols

	* configure.ac: updated configure.ac logic to deal with /lib ->
	  /usr/lib and /lib64 -> /usr/lib64 as with Fedora 17 - the
	  previous logic was too clever for its own good

2012-06-12 07:37  david_nichols

	* src/QC_SFTPClient.qpp: fixed warnings

2012-06-11 17:08  david_nichols

	* src/SFTPClient.h, src/ssh2-module.cpp: build fixes

2012-06-08 14:58  david_nichols

	* src/single-compilation-unit.cpp: checked in missing file

2012-06-08 12:59  david_nichols

	* Makefile.am, configure.ac: minor updates

2012-06-08 10:49  david_nichols

	* docs/doxygen-api-filter.q: removed unneeded/deprecated script

2012-06-08 09:33  david_nichols

	* src/Makefile.am: enabled single compilation unit build

2012-06-08 09:23  david_nichols

	* Makefile.am, docs/mainpage.doxygen.tmpl, qore-ssh2-module.spec,
	  src/Makefile.am, src/QC_SFTPClient.cpp, src/QC_SFTPClient.qpp,
	  src/QC_SSH2Client.qpp: completed migration to qpp; doc updates

2012-06-08 08:51  david_nichols

	* src/QC_SSH2Client.qpp, src/SSH2Channel.h, src/SSH2Client.h,
	  src/ssh2-module.cpp: fixed bugs in last commit

2012-06-08 08:29  david_nichols

	* Makefile.am, configure.ac, docs/doxyfile.tmpl,
	  docs/mainpage.doxygen.tmpl, src/Makefile.am,
	  src/QC_SFTPClient.cc, src/QC_SFTPClient.cpp, src/QC_SSH2Base.cc,
	  src/QC_SSH2Base.h, src/QC_SSH2Base.qpp, src/QC_SSH2Channel.cc,
	  src/QC_SSH2Channel.qpp, src/QC_SSH2Client.cc,
	  src/QC_SSH2Client.qpp, src/SFTPClient.cc, src/SFTPClient.cpp,
	  src/SSH2Channel.cc, src/SSH2Channel.cpp, src/SSH2Client.cc,
	  src/SSH2Client.cpp, src/ssh2-module.cc, src/ssh2-module.cpp: ssh2
	  module partially converted to qpp

2012-05-22 13:38  david_nichols


src/QC_SSH2Base.cc
/* -*- mode: c++; indent-tabs-mode: nil -*- */
/*
  QC_SSH2Base.h

  libssh2 ssh2 client integration in Qore

  Qore Programming Language

  Copyright 2009 Wolfgang Ritzinger

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

//! @file SSH2Base.qc defines the SSH2Base class

#include "QC_SSH2Base.h"
#include "SSH2Client.h"

qore_classid_t CID_SSH2_BASE;

static const char *SSH2_CONNECTED = "SSH2-CONNECTED";

//! namespace for the SSH2 module
/**# namespace SSH2 {
*/
//! base class for SFTPClient and SSH2Client
/** The SSH2Base class provides common methods to the SSH2Client and SFTPClient classes
 */
/**# class SSH2Base {
public:
 */

//! Throws an exception; the constructor cannot be called manually
/** Throws an exception if called directly; this class cannot be instantiated directly
    @throw SSH2BASE-CONSTRUCTOR-ERROR this class is an abstract class and cannot be instantiated directly or directly inherited by a user-defined class
 */
//# constructor() {}
void SSH2BASE_constructor(QoreObject *self, const QoreListNode *params, ExceptionSink *xsink) {
   xsink->raiseException("SSH2BASE-CONSTRUCTOR-ERROR", "this class is an abstract class and cannot be instantiated directly or directly inherited by a user-defined class");
}

//! connect to remote system
/** Connects to the remote system; if a connection is already established, then it is disconnected first

    @param $timeout an integer giving a timeout in milliseconds or a relative date/time value (ex: \c 15s for 15 seconds)

    @throw SOCKET-CONNECT-ERROR error establishing socket connection (no listener, port blocked, etc); timeout establishing socket connection
    @throw SSH2CLIENT-CONNECT-ERROR no user name set; ssh2 or libssh2 error
    @throw SSH2-ERROR error initializing or establishing ssh2 session
    @throw SSH2CLIENT-AUTH-ERROR no proper authentication method found
    @throw SFTPCLIENT-CONNECT-ERROR error initializing sftp session or getting remote path (exception only possible when called from an SFTPClient object)

    @par Example:
    @code $sftpclient.connect(30s); @endcode
 */
//# nothing connect(timeout $timeout = -1) {}
static AbstractQoreNode *SSH2BASE_connect(QoreObject *self, SSH2Client *myself, const QoreListNode *params, ExceptionSink *xsink) {
   myself->connect(getMsMinusOneInt(get_param(params, 0)), xsink);
   return 0;
}

//! Disconnects from the remote system; throws an exception if the object is not currently connected
/** @throw SSH2CLIENT-NOT-CONNECTED the client is not connected

    @par Example:
    @code $sftpclient.disconnect(); @endcode
 */
//# nothing disconnect() {}
static AbstractQoreNode *SSH2BASE_disconnect(QoreObject *self, SSH2Client *myself, const QoreListNode *params, ExceptionSink *xsink) {
   myself->disconnect(0, xsink);
   return 0;
}

//! Sets the user name for the next connection; can only be called when a connection is not established, otherwise an exception is thrown
/** @param $user the user name to set for the next connection

    @throw SSH2-CONNECTED this method cannot be called when a connection is established

    @par Example:
    @code $sftpclient.setUser("username"); @endcode
 */
//# nothing setUser(string $user) {}
static AbstractQoreNode *SSH2BASE_setUser(QoreObject *self, SSH2Client *myself, const QoreListNode *params, ExceptionSink *xsink) {
   const QoreStringNode *p0 = HARD_QORE_STRING(params, 0);

   if (myself->setUser(p0->getBuffer()))
      xsink->raiseException(SSH2_CONNECTED, "usage of SSH2Base::setUser() is not allowed when connected");

   return 0;
}

//! Sets the password for the next connection; can only be called when a connection is not established, otherwise an exception is thrown
/** @param $pass the password to use for the next connection

    @throw SSH2-CONNECTED this method cannot be called when a connection is established

    @par Example:
    @code $sftpclient.setPassword("pass"); @endcode
 */
//# nothing setPassword(string $pass) {}
static AbstractQoreNode *SSH2BASE_setPassword(QoreObject *self, SSH2Client *myself, const QoreListNode *params, ExceptionSink *xsink) {
   const QoreStringNode *p0 = HARD_QORE_STRING(params, 0);

   if (myself->setPassword(p0->getBuffer()))
      xsink->raiseException(SSH2_CONNECTED, "usage of SSH2Base::setPassword() is not allowed when connected");

   return 0;
}

//! Sets path to the private key and optionally the public key to use for the next connection; can only be called when a connection is not established, otherwise an exception is thrown
/** @param $priv_key the path to the private key file to use for the next connection
    @param $pub_key optional: the path to the public key file to use for the next connection

    @throw SSH2-CONNECTED this method cannot be called when a connection is established

    @par Example:
    @code $sftpclient.setKeys($ENV.HOME + "/.ssh/id_rsa", $ENV.HOME + "/.ssh/id_rsa.pub"); @endcode
 */
//# nothing setKeys(string $priv_key, *string $pub_key) {}
static AbstractQoreNode *SSH2BASE_setKeys(QoreObject *self, SSH2Client *myself, const QoreListNode *params, ExceptionSink *xsink) {
   const QoreStringNode *p0 = HARD_QORE_STRING(params, 0);
   const QoreStringNode *p1 = test_string_param(params, 1);

   if (myself->setKeys(p0->getBuffer(), p1 ? p1->getBuffer() : 0))
      xsink->raiseException(SSH2_CONNECTED, "usage of SSH2Base::setKeys() is not allowed when connected");

   return 0;
}

/**# };
};
*/

QoreClass *initSSH2BaseClass() {
   QORE_TRACE("initSSH2Base()");

   QoreClass *QC_SSH2_BASE = new QoreClass("SSH2Base", QDOM_NETWORK);
   CID_SSH2_BASE = QC_SSH2_BASE->getID();
   QC_SSH2_BASE->setConstructor(SSH2BASE_constructor);

   // SSH2Base::connect(softint $timeout_ms = -1) returns nothing
   QC_SSH2_BASE->addMethodExtended("connect",      (q_method_t)SSH2BASE_connect, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo, 1, softBigIntTypeInfo, new QoreBigIntNode(-1));
   // SSH2Base::connect(date $timeout) returns nothing
   QC_SSH2_BASE->addMethodExtended("connect",      (q_method_t)SSH2BASE_connect, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo, 1, dateTypeInfo, QORE_PARAM_NO_ARG);

   // SSH2Base::disconnect() returns nothing
   QC_SSH2_BASE->addMethodExtended("disconnect",   (q_method_t)SSH2BASE_disconnect, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo);

   // SSH2Base::setUser(string $user) returns nothing
   QC_SSH2_BASE->addMethodExtended("setUser",      (q_method_t)SSH2BASE_setUser, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo, 1, stringTypeInfo, QORE_PARAM_NO_ARG);

   // SSH2Base::setPassword(string $pass) returns nothing
   QC_SSH2_BASE->addMethodExtended("setPassword",  (q_method_t)SSH2BASE_setPassword, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo, 1, stringTypeInfo, QORE_PARAM_NO_ARG);

   // SSH2Base::setKeys(string $priv_key) returns nothing
   // SSH2Base::setKeys(string $priv_key, string $pub_key) returns nothing
   QC_SSH2_BASE->addMethodExtended("setKeys",      (q_method_t)SSH2BASE_setKeys, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo, 1, stringTypeInfo, QORE_PARAM_NO_ARG);
   QC_SSH2_BASE->addMethodExtended("setKeys",      (q_method_t)SSH2BASE_setKeys, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo, 2, stringTypeInfo, QORE_PARAM_NO_ARG, stringTypeInfo, QORE_PARAM_NO_ARG);

   return QC_SSH2_BASE;
}
